lib/galaxy/model/store/__init__.py

```diff
@@ -182,6 +183,30 @@ def perform_import(self, history=None, new_history=False, job=None):
         self._import_implicit_collection_jobs(object_import_tracker)
         self._flush()
 
+    def _attach_dataset_hashes(self, dataset_or_file_attrs, dataset_instance):
+        if "hashes" in dataset_or_file_attrs:
+            for hash_attrs in dataset_or_file_attrs["hashes"]:
+                hash_obj = model.DatasetHash()
+                hash_obj.hash_value = hash_attrs["hash_value"]
+                hash_obj.hash_function = hash_attrs["hash_function"]
+                hash_obj.extra_files_path = hash_attrs["extra_files_path"]
+                dataset_instance.dataset.hashes.append(hash_obj)
+
+    def _attach_dataset_sources(self, dataset_or_file_attrs, dataset_instance):
+        if "sources" in dataset_or_file_attrs:
+            for source_attrs in dataset_or_file_attrs["sources"]:
+                source_obj = model.DatasetSource()
+                source_obj.source_uri = source_attrs["source_uri"]
+                source_obj.transform = source_attrs["transform"]
+                source_obj.extra_files_path = source_attrs["extra_files_path"]
+                for hash_attrs in source_attrs["hashes"]:
+                    hash_obj = model.DatasetSourceHash()
+                    hash_obj.hash_value = hash_attrs["hash_value"]
```

Comment:
Do we need to verify these if they come from another Galaxy instance? It would slow down import, but if we want to rely on these for deduplication or as a sort "this is safe, it was produced in Galaxy" kind of check

Reply:
By verify - do you mean checking these hashes? It seems like that is a decision we should make it once we're actually using this metadata to re-populate datasets - it is all just provenance currently.

Reply:
> By verify - do you mean checking these hashes?

Yes.

> It seems like that is a decision we should make it once we're actually using this metadata to re-populate datasets - it is all just provenance currently.

I'm confused by this, doesn't the unit test show that the metadata is attached to the datasets ? Which, if true, means that at this point we don't know if the hash has been tampered with or not.

If this is just for provenance we could limit this to the case where no data is being imported ? Or make it explicit somewhere that we didn't check the hash ?

Reply:
I think you can already use the data fetch API to put in bad hashes and sources for an uploaded file. This isn't meant to be a security feature. 

> Or make it explicit somewhere that we didn't check the hash ?

I mean - I would do the inverse... and maybe add a column to database when the data has been verified?

Reply:
Alright, that sounds good. I had assumed Galaxy calculates the hash, not that we can pass it in on upload.

Reply:
Ahh - sure. Thanks for the merge.